src/utils/migration.ts

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2018 - 2022 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { encrypt } from '@alephium/sdk/dist/lib/password-crypto'
+
+export const latestUserDataVersion = '2022-05-27T12:00:00Z'
+
+//
+// Arguments are essentially dependencies of the migrations.
+// Unfortunately mnemonics are a dependency due to being needed to encrypt address
+// metadata, so migration must absolutely occur after login at the earliest.
+//
+// Future dependencies must be explained as they are added.
+//
+export const migrateUserData = (mnemonic: string) => {
+  console.log('🚚 Migrating user data')
+  _20220527_120000(mnemonic)
+}
+
+const _20220527_120000 = (mnemonic: string) => {
+  const addressMetadata = Object.entries(localStorage).filter(([key, value]) => /^addresses-metadata/.test(key))
+  for (const [key, json] of addressMetadata) {
+    const addressSettingsList = JSON.parse(json)
+
+    //
+    // Means the old format is being used, which is not encrypted.
+    // The data structure can be serialized and then encrypted.
+    // We can also take this opportunity to start versioning our data.
+    //
+    if (Array.isArray(addressSettingsList)) {
+      localStorage.setItem(
+        key,
+        JSON.stringify({
+          version: '2022-05-27T12:00:00Z',
+          encryptedSettings: encrypt(mnemonic, JSON.stringify(addressSettingsList))
+        })
+      )
+    }
+  }
+}
```

Comment:
The migration function does not work properly. I had 10 accounts set up, and after logging in to one of them (called `Ilias Testnet`) I noticed that a new entry in localStorage was created with the key:
```
addresses-metadata-b2189fac31a86e96d637b492378fbda08f51bfebfa9e3b2f95f5d36246b75d85c2a9eea9e545c2bfd85dbf4d5b7deae4986dad6e18109c8842790d8bd695bc87
```
and value
```
{"version":"2022-05-27T12:00:00Z","encryptedSettings":"{\"salt\":\"87916ec83063b759891903b28b538403f3c42a633067302bd972c986010aef25346920675f38b4f23432da14c0c2c95eec50ea3c4342fc96a0d350781a419834\",\"iv\":\"9979900b2bc7327f3d9f623d783aa04a9ec30126b6c909f5372f169119645a191c3c9988ba6f7436eed1a9a0840dae10a080c981bc9832e099a765406fbc04c1\",\"encrypted\":\"956a0f2a62f03021f21d482b825f463d6652eb4b3ca30bece74e9b0645e3148ec7076aa62409303e3148f6\",\"version\":1}"}
```
but the entry with key
```
addresses-metadata-Ilias Testnet
```
still exists, but it's value has changed to:
```
{"version":"2022-05-27T12:00:00Z","encryptedSettings":"{\"salt\":\"b1d127f6ec4bc237cea3d3620c4495e01c4d84d47498cf0f026c6bd6e8c751b0fdd2444f5c54c187fe13403be6cfe51d047b158a66a5007606029586c9b55e0c\",\"iv\":\"0a0c88b067a00dbf60fec7c7bdd5bbd1060443011afecd56a4fe2bd56ea9496f5fbee834b4ed2924b0bc14174fb0c8a1f629af0db70106f4f6eca12e4f77a9c8\",\"encrypted\":\"318f9e0cf4326b5a0d173b121d6e8ddab828e4ed3f12ab2933d70ec9240713499a48d8b0393590dca88320\",\"version\":1}"}
```

Secondly, I lost all my metadata. No matter which account I login, there's no metadata there (no address labels, no additional addresses, nothing).

Third, I realize that all metadata of all wallets are encrypted with the mnemonic of the account I logged in with, not just the metadata of the account I logged in with. That doesn't make sense.

Reply:
>  looks like it wasn't tested

It was tested but clearly not enough; I created wallets before the change, tested after the change, and I was able to see the old metadata...

> still exists

The old key shouldn't exist anymore; bug

> Each account metadata should be encrypted with its own mnemonic, no?

Yeah, nice catch, and is most likely why you lost the metadata originally... Rather than iterate over all address metadata properties I should only migrate based on the account which is selected

---

Thank you for testing; features like this need to really be scrutinized so I really appreciate it 🙂